lib/drivers/docker/index.js

```diff
@@ -0,0 +1,294 @@
+    case 'hard':
+    default:
+      if (instance.current) {
+        return instance.current.kill('SIGTERM', reportExit);
+      } else {
+        return cb(Error('No such instance: ' + id));
+      }
+  }
+  function reportExit(err, res) {
+    debug('stopped container for instance %j:', id, err, res);
+    cb(err, res);
+  }
+};
+
+DockerDriver.prototype._onContainerStopped = function(id, app) {
+  console.error('container stream stopped for some reason!', id, app);
+};
+
+DockerDriver.prototype.dumpInstanceLog = function(id) {
+  debug('DockerDriver.dumpInstanceLog(%j)', id);
+  var instance = this._instance(id);
+  if (!instance.current) {
+    debug('no instance to dump logs for: %j', id);
+    return null;
+  }
+  var logDump = instance.log.duplicate().toString();
+  instance.log.consume(logDump.length);
+  return logDump;
+};
+
+DockerDriver.prototype.updateInstanceEnv = function(id, env, cb) {
+  debug('DockerDriver.updateInstanceEnv(%j)', id);
+  var self = this;
+  var current = this._instance(id).current;
+  if (!current) {
+    debug('no instance to update env of: %j', id);
+    return setImmediate(cb);
+  }
+  // Environment variables can only really have string values, so make sure we
+  // are consistent with that so our comparisons are reliable.
+  env = _.mapValues(env, String);
+  debug('updating instance %j env: %j', id, env);
+  if (_.isEqual(current.env, env)) {
+    debug('env unchanged, skipping restart: %j', id);
+    return cb && setImmediate(cb);
+  } else {
+    debug('env changed, restarting %j: %j => %j', id, current.env, env);
+    // env is alredy a copy
+    current.env = env;
+    return self.requestOfInstance(id, {cmd: 'env-set', env: env}, function() {
+      self.requestOfInstance(id, {cmd: 'restart'}, function(rsp) {
+        cb(rsp && rsp.error, rsp);
+      });
+    });
+  }
+};
+
+DockerDriver.prototype.requestOfInstance = function(id, req, cb) {
+  debug('DockerDriver.requestOfInstance(%j)', id);
+  var current = this._instance(id).current;
+  if (current) {
+    debug('requesting of %j: %j', id, req);
+    return current.request(req, cb);
+  } else if (cb) {
+    setImmediate(cb({error: 'Cannot send request to instance'}));
+  }
+};
+
+DockerDriver.prototype.start = function(instanceMetas, cb) {
+  debug('DockerDriver.start(%j, %s)', instanceMetas, cb.name);
+  // TODO(rmg):
+  //   1. Get list of currently running docker containers.
+  //   2. Compare containers to instanceMetas
+  //   3. Create internal instances from matching containers
+  //   4. Start new containers from missing instances
+  var self = this;
+  this.docker.info(function(err, info) {
+    self.dockerInfo = info;
+    self.CPUS = info.NCPU;
+    debug('Connected to docker:', err, info);
+    cb(err);
+  });
+};
+
+DockerDriver.prototype.stop = function(cb) {
+  var self = this;
+
+  async.each(Object.keys(this.instances), stopInstanceById, cb);
+
+  function stopInstanceById(id, next) {
+    self.stopInstance(id, 'hard', next);
```

Comment:
if this can fail with an error, then you should wrap the callback.. otherwise the first docker image that doesn't stop will block the stop of all others.

Also, I commented out `stopWhenDone()` in index.js, since direct children already auto-exit when their parent exits.

But.. I suspect it should come back, so the driver can be stopped, and it can have a chance to stop all the containers it started. Otherwise, `ctrl-C` of sl-pm will always leave containers running.